drivers/serial/uart_xlnx_ps.c

```diff
@@ -313,6 +385,446 @@ static void uart_xlnx_ps_poll_out(struct device *dev, unsigned char c)
+ *        enum value.
+ *
+ * Converts a bit mask representing the UART's stop bit setting within
+ * the UART's Mode Register into a value of an enumeration type provided
+ * by the UART driver API.
+ *
+ * @param mode_reg The current Mode Register contents from which the
+ *                 stop bit setting shall be extracted.
+ *
+ * @return The current stop bit setting mapped to the UART driver API's
+ *         enum type.
+ */
+static inline enum uart_config_stop_bits uart_xlnx_ps_ll2cfg_stopbits(
+	u32_t mode_reg)
+{
+	/*
+	 * Obtain the current stop bit configuration from the mode register's
+	 * bits [7..6] (NBSTOP):
+	 *  00b : 1 stop bit -> reset value
+	 *  01b : 1.5 stop bits
+	 *  10b : 2 stop bits
+	 *  11b : reserved
+	 */
+
+	switch ((mode_reg & XUARTPS_MR_STOPMODE_MASK)) {
+	case XUARTPS_MR_STOPMODE_1_BIT:
+	default:
+		return UART_CFG_STOP_BITS_1;
+	case XUARTPS_MR_STOPMODE_1_5_BIT:
+		return UART_CFG_STOP_BITS_1_5;
+	case XUARTPS_MR_STOPMODE_2_BIT:
+		return UART_CFG_STOP_BITS_2;
+	}
+}
+
+/**
+ * @brief Converts a Mode Register bit mask to a data bit configuration
+ *        enum value.
+ *
+ * Converts a bit mask representing the UART's data bit setting within
+ * the UART's Mode Register into a value of an enumeration type provided
+ * by the UART driver API.
+ *
+ * @param mode_reg The current Mode Register contents from which the
+ *                 data bit setting shall be extracted.
+ *
+ * @return The current data bit setting mapped to the UART driver API's
+ *         enum type.
+ */
+static inline enum uart_config_data_bits uart_xlnx_ps_ll2cfg_databits(
+	u32_t mode_reg)
+{
+	/*
+	 * Obtain the current data bit configuration from the mode register's
+	 * bits [2..1] (CHRL):
+	 *  0xb : 8 data bits -> reset value
+	 *  10b : 7 data bits
+	 *  11b : 6 data bits
+	 */
+
+	switch ((mode_reg & XUARTPS_MR_CHARLEN_MASK)) {
+	case XUARTPS_MR_CHARLEN_8_BIT:
+	default:
+		return UART_CFG_DATA_BITS_8;
+	case XUARTPS_MR_CHARLEN_7_BIT:
+		return UART_CFG_DATA_BITS_7;
+	case XUARTPS_MR_CHARLEN_6_BIT:
+		return UART_CFG_DATA_BITS_6;
+	}
+}
+
+/**
+ * @brief Converts a Modem Control Register bit mask to a flow control
+ *        configuration enum value.
+ *
+ * Converts a bit mask representing the UART's flow control setting within
+ * the UART's Modem Control Register into a value of an enumeration type
+ * provided by the UART driver API.
+ *
+ * @param modemcr_reg The current Modem Control Register contents from
+ *                    which the parity setting shall be extracted.
+ *
+ * @return The current flow control setting mapped to the UART driver API's
+ *         enum type.
+ */
+static inline enum uart_config_flow_control uart_xlnx_ps_ll2cfg_hwctrl(
+	u32_t modemcr_reg)
+{
+	/*
+	 * Obtain the current flow control configuration from the modem
+	 * control register's bit [5] (FCM):
+	 *  0b : no flow control -> reset value
+	 *  1b : RTS/CTS
+	 */
+
+	if ((modemcr_reg & XUARTPS_MODEMCR_FCM_MASK)
```

Comment:
I'm wondering if this (and similar) needs to have `&&` in the if and else if statement logic?

Reply:
No, this is effectively masking the `MODEMCR` reg `FCM` field to check if the field is set to a specific value. We **must** use bitwise AND.